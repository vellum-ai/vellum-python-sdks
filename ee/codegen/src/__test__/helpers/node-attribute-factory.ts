import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
