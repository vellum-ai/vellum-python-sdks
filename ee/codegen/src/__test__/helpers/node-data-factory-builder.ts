import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
