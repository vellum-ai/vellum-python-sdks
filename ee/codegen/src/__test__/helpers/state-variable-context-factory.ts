import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
