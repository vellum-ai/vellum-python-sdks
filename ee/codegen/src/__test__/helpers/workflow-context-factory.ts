import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
