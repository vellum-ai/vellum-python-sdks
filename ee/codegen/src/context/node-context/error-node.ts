import { VellumVariableType } from "vellum-ai/api";

import { BaseNodeContext } from "./base";

import { PortContext } from "src/context/port-context";
import { ErrorNode } from "src/types/vellum";

export class ErrorNodeContext extends BaseNodeContext<ErrorNode> {
  baseNodeClassName = "ErrorNode";
  baseNodeDisplayClassName = "BaseErrorNodeDisplay";
  isCore = true;

  getNodeOutputNamesById(): Record<string, string> {
    return {};
  }

  getNodeOutputTypesById(): Record<string, VellumVariableType> {
    return {};
  }

  createPortContexts(): PortContext[] {
    return [];
  }
}
