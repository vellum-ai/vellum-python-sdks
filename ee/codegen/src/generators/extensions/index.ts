/**
 * This directory contains custom AST Nodes and classes that should one day be contributed to Fern's python-ast package.
 * We do this to unblock critical functionality for our users to avoid forking the package or being blocked on the
 * upstream package being updated.
 *
 * The process usually entails:
 * 1. Creating a new file in this directory that supports the new functionality you need.
 * 2. Exporting the new class from this file to use in the rest of codegen.
 * 3. In parallel, create a PR at https://github.com/fern-api/fern/tree/main/generators/python-v2/ast
 * 4. Once the PR is merged, update the dependency from our dedicated branch: https://github.com/fern-api/fern/pull/5100
 * 5. Remove the custom code in this directory and use the new AST Node from the upstream package.
 */

export * from "./protected-python-file";
