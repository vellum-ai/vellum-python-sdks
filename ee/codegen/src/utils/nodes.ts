import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
