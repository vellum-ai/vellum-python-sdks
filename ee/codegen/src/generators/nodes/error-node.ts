import { python } from "@fern-api/python-ast";
import { Field } from "@fern-api/python-ast/Field";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { ErrorNodeContext } from "src/context/node-context/error-node";
import { BaseNode } from "src/generators/nodes/bases/base";
import { ErrorNode as ErrorNodeType } from "src/types/vellum";

export class ErrorNode extends BaseNode<ErrorNodeType, ErrorNodeContext> {
  getNodeClassBodyStatements(): AstNode[] {
    const bodyStatements: AstNode[] = [];
    const errorSourceInputId = this.getNodeInputByName("error_source_input_id");

    if (errorSourceInputId) {
      bodyStatements.push(
        python.field({
          name: "error",
          initializer: errorSourceInputId,
        })
      );
    }

    return bodyStatements;
  }

  getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    if (this.nodeData.data.name) {
      // DEPRECATED: To be removed in the 0.15.0 release
      statements.push(
        python.field({
          name: "name",
          initializer: python.TypeInstantiation.str(this.nodeData.data.name),
        })
      );
    }

    statements.push(
      python.field({
        name: "node_id",
        initializer: python.TypeInstantiation.uuid(this.nodeData.id),
      })
    );
    statements.push(
      python.field({
        name: "label",
        initializer: python.TypeInstantiation.str(this.nodeData.data.label),
      })
    );

    if (this.nodeData.data.errorOutputId) {
      // DEPRECATED: To be removed in the 0.15.0 release
      statements.push(
        python.field({
          name: "error_output_id",
          initializer: python.TypeInstantiation.uuid(
            this.nodeData.data.errorOutputId
          ),
        })
      );
    }

    statements.push(
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    return statements;
  }

  protected getOutputDisplay(): Field | undefined {
    return undefined;
  }

  getErrorOutputId(): string | undefined {
    return undefined;
  }
}
