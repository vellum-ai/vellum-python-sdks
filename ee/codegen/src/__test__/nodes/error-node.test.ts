import { beforeEach } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { errorNodeDataFactory } from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { ErrorNodeContext } from "src/context/node-context/error-node";
import { Writer } from "src/generators/extensions";
import { ErrorNode } from "src/generators/nodes/error-node";

describe("ErrorNode", () => {
  let workflowContext: WorkflowContext;
  let node: ErrorNode;
  let writer: Writer;

  beforeEach(async () => {
    writer = new Writer();
  });

  describe("basic", () => {
    beforeEach(async () => {
      workflowContext = workflowContextFactory();
      writer = new Writer();

      const nodeData = errorNodeDataFactory();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as ErrorNodeContext;

      node = new ErrorNode({
        workflowContext,
        nodeContext,
      });
    });

    it(`getNodeFile`, async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it(`getNodeDisplayFile`, async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("should codegen successfully without error source inputs", () => {
    beforeEach(async () => {
      workflowContext = workflowContextFactory({ strict: false });

      const nodeData = errorNodeDataFactory({
        errorSourceInputs: [],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as ErrorNodeContext;

      node = new ErrorNode({
        workflowContext,
        nodeContext,
      });
    });

    it(`getNodeFile`, async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it(`getNodeDisplayFile`, async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
