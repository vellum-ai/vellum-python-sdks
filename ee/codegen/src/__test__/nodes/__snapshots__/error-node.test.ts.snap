// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorNode > basic > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseErrorNodeDisplay
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.error_node import ErrorNode


class ErrorNodeDisplay(BaseErrorNodeDisplay[ErrorNode]):
    name = "error-node"
    node_id = UUID("2cd960a3-cb8a-43ed-9e3f-f003fc480951")
    label = "Error Node"
    error_output_id = UUID("69250713-617d-42a4-9326-456c70d0ef20")
    target_handle_id = UUID("370d712d-3369-424e-bcf7-f4da1aef3928")
    node_input_ids_by_name = {
        "error_source_input_id": UUID("d2287fee-98fb-421c-9464-e54d8f70f046")
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`ErrorNode > basic > getNodeFile 1`] = `
"from vellum import VellumError
from vellum.workflows.nodes.displayable import ErrorNode as BaseErrorNode


class ErrorNode(BaseErrorNode):
    error = VellumError(message="Something went wrong!", code="USER_DEFINED_ERROR")
"
`;

exports[`ErrorNode > should codegen successfully without error source inputs > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseErrorNodeDisplay
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.error_node import ErrorNode


class ErrorNodeDisplay(BaseErrorNodeDisplay[ErrorNode]):
    name = "error-node"
    node_id = UUID("2cd960a3-cb8a-43ed-9e3f-f003fc480951")
    label = "Error Node"
    error_output_id = UUID("69250713-617d-42a4-9326-456c70d0ef20")
    target_handle_id = UUID("370d712d-3369-424e-bcf7-f4da1aef3928")
    node_input_ids_by_name = {}
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`ErrorNode > should codegen successfully without error source inputs > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ErrorNode as BaseErrorNode


class ErrorNode(BaseErrorNode):
    pass
"
`;
